### ISSUE TITLE
Slow down in npvalue and how to do BPTT

### ISSUE THREAD

**Issue by emanjavacas** [clab/dynet#1410]
Hi there,

I am using `prev_hidden = dynet.inputTensor([h.npvalue() for h in state.s()])` to keep the hidden state of the previous sentence. Then I use `self.builder.initial_state(prev_hidden)` on the next sequence to initialize with the previous hidden state and I am observing that everything slows down by a factor of 3. Not sure why that operation would cause such a slow down.

A related question would be how to actually backpropagate through a number of consecutive sequences. I can imagine that since we are extracting the tensor from the expression, the computational graph is also cut from its previous history. Still I would like to backpropagate through a number of consecutive sentences to implement a kind of BPTT at the level of sequences.

Any help will be very much welcomed! Thanks!

**Comment by pmichel31415** [comment]
So there are 2 things here:

- You should call `self.builder.initial_state(prev_hidden)` if `prev_hidden` is a list of `dy.Expression`s vs `self.builder.initial_state_from_raw_vectors(prev_hidden)` if `prev_hidden` is a list of numpy arrays
- Converting from Expressions to Numpy arrays is coslty because it requires copying memory. Instead you could: 
  1. `self.builder.initial_state(state.s())` (make sure you don't call renew_cg between the two calls to `initial_state` otherwise `state.s()` will return stale expressions)
  2. Continue using `state` to process the next sentence

LEt me know if you have more questions.
You can reuse the

**Comment by emanjavacas** [comment]
Thanks, in the end I moved that implementation to Pytorch, but it'd be nice to know for the next time around.

**Comment by emanjavacas** [comment]
Hi, I am looking at your suggestion (i) and I think I am not getting it - I thought I did. What I am trying to do is to initialize the hidden state at sentence `i` with the last hidden state at sentence `i-1`. But I do not want to backpropagate gradients into sentence `i`. The problem I have there is that `state.s()` wouldn't work, because the graph has been renewed and `initial_state(prev_hidden)` is slow because `prev_hidden` involves a call to `np.value()`. Is there a way of initializing the state of a new computational graph with a tensor from previous one without having to copy memory?

**Comment by FilippoC** [comment]
If you were a C++ user, I would suggest you to try to use a model parameter that is used to save the hidden layer between computational graphs:
- in sentence i-1, you save the hidden layer using `TensorTool::copy_elements(*previous_hidden_layer.values(), expr.value())`
- in sentence i you include the hidden layer with `expr = dynet::const_parameter(cg, previous_hidden_layer)`

I don't know if this actually works neither if you have access to underlying tensors in python, but it may be worth a try.

**Comment by emanjavacas** [comment]
Yeah, I am looking for something like that but unfortunately I am not using C++... hopefully some of the pythonists around has a suggestion for doing it in Python. Thanks for the help!
